apiserver/facades/client/charms/client.go

```diff
@@ -380,6 +412,77 @@ func (a *API) addCharmWithAuthorization(args params.AddCharmWithAuth) (params.Ch
 	}, nil
 }
 
+func (a *API) queueAsyncCharmDownload(args params.AddCharmWithAuth) (corecharm.Origin, error) {
+	charmURL, err := charm.ParseURL(args.URL)
+	if err != nil {
+		return corecharm.Origin{}, err
+	}
+
+	// Check if a charm doc already exists for this charm URL. If so, the
+	// charm has already been queued for download so this is a no-op. We
+	// still need to return back the recorded origin from where the charm
+	// is to be downloaded.
+	if _, err := a.backendState.Charm(charmURL); err == nil {
```

Comment:
Hmm... Looks like I don't fully understand how the process works :-( 

AFAICT from the original version of the code, charm URLs (the ID in charm documents) are supposed to be unique; the code explicitly checks for the presence of a charm doc and treats AddCharm as a no-op. However, in such a case, the original version of the code would also resolve the origin **again** and return it to the client.

I believe this is wrong. Resolving the origin again might yield a different origin (because the developer released a new version in the meantime) which means that Juju deploys blob B (fetched after resolving the origin to **R0**) but we may record a _potentially different_ origin R' against the application doc (via the subsequent deploy steps).

Reply:
So this is known and probably what epochs are intended to solve...?

Reply:
So, is the right ™️ solution here to preserve the original behavior and always resolve the origin?

Reply:
Ok, after chatting with Simon, I know have a better understanding about how charmhub revisions can point to the same blob but in different channels. The above bit was probably correct until https://github.com/juju/juju/pull/13206 landed (the actual introduction of `UploadedCharmOrigin` landed before this one as part of the refactoring work in https://github.com/juju/juju/pull/13191. 

So I will rework the PR to drop `UploadedCharmOrigin` call and ensure we always resolve the origin when the charm is already downloaded. On the bright side, once we transition to server-side bundle expansion, the controller will resolve the origin just once and this re-resolve step will be redundant.